git_repo/services/ext/gogs.py

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python
+
+import logging
+log = logging.getLogger('git_repo.gogs')
+
+from ..service import register_target, RepositoryService, os
+from ...exceptions import ResourceError, ResourceExistsError, ResourceNotFoundError
+
+import requests
+from urllib.parse import urlparse, urlunparse
+import functools
+
+from git import config as git_config
+from git.exc import GitCommandError
+
+@register_target('gg', 'gogs')
+class GoGSService(RepositoryService):
+    fqdn = 'try.gogs.io'
+
+    @classmethod
+    def _url_parse(cls, url):
+        if '://' not in url:
+            url = 'https://'+url
+        parse = urlparse(url)
+        url_base = urlunparse((parse.scheme, parse.netloc)+('',)*4)
+        fqdn = parse.hostname
+        return url_base, fqdn
+
+    @property
+    def url_ro(self):
+        return self.url_base
+
+    @property
+    def url_rw(self):
+        url = self.ssh_url
+        if '@' in url:
+            return url
+        return '@'.join([self.git_user, url])
+
+    def url_api(self, rest):
+        return '{}/api/v1/{}'.format(self.url_base, rest)
+
+    @classmethod
+    def get_auth_token(cls, login, password, prompt=None):
+        import platform
+        name = 'git-repo2 token used on {}'.format(platform.node())
+        if '/' in login:
+            url, login = login.rsplit('/', 1)
+        else:
+            url = input('URL [{}]> '.format(cls.fqdn)) or cls.fqdn
+        url_base, fqdn = cls._url_parse(url)
+        url_api = functools.partial('{}/api/v1/{}'.format, url_base)
+        r = requests.get(url_api('users/{}/tokens'.format(login)), auth=(login, password), verify=False)
+        r.raise_for_status()
+        tokens = r.json()
+        tokens = dict((o['name'], o['sha1']) for o in tokens)
+        if name in tokens:
+            return tokens[name]
+        if 'git-repo2 token' in tokens:
+            return tokens['git-repo2 token']
+        r = requests.post(url_api('users/{}/tokens'.format(login)), auth=(login, password), json=dict(name=name), verify=False)
+        r.raise_for_status()
+        token = r.json()
+        return token['sha1']
+
+    @property
+    def user(self):
+        r = self.session.get(self.url_api('user'))
+        r.raise_for_status()
+        user = r.json()
+        return user['username']
+
+    def orgs(self):
+        r  = self.session.get(self.url_api('user/orgs'))
+        r.raise_for_status()
+        return [o['username'] for o in r.json()]
+
+    def connect(self):
+        self.url_base, self.fqdn = self._url_parse(self.fqdn)
+        verify = self.fqdn == 'try.gogs.io'
+        #verify = True
+        verify = self.config.get('verify', ['no','yes'][verify])
+        if verify.lower().strip() in ('0','no','false',''):
+            verify = False
+        elif verify.lower().strip() in ('1','yes','true'):
+            verify = True
```

Comment:
This feature is being implemented [in another pull request](https://github.com/guyzmo/git-repo/pull/76/files), which is ready and will be released along the next release (which is due relatively soon, it's mostly depending on Gitlab to merge a patch I made).

So I'm merging that PR into devel (I actually thought I already did), so you can rebase your branch and use `self.insecure` instead of verify.